### PR TITLE
Reduce CI scope on pushes and pull requests

### LINKED ISSUE #1417: Narrow the scope of CI on push to non-integration branches
### Proposed Changes

Currently a push to any branch triggers the full CI suite including the full matrix of Python versions and Postgres/MySQL tests. Propose changing things so that while pushes to integration branches (`main`, `develop`, `next`) will continue to run the full CI suite, pushes to any other branch should run a much smaller set of tests, possibly just linters + 1 unittest suite.

### Justification

CI shouldn't be used as a substitute for running tests locally, and the current setup results in a lot of load on GitHub Actions when multiple developers are pushing commits for multiple feature branches.

Additionally, once a PR is open for a given branch, the current setup results in the full CI suite being run twice on each push (once for the push, once for the PR update), which is excessive and redundant.

### PR DESCRIPTION
**Pull request by glennmatthews**
Fixes #1417 
# What's Changed

Borrowing patterns from #1184, this basically splits `ci.yml` into two files:
- `ci_integration.yml` - full CI suite, essentially identical steps to the old `ci.yml`, now only runs on pushes to `main`, `develop`, and `next` (not on pushes to individual feature branches - developers working in a feature branch should run tests locally as appropriate)
- `ci_pullrequest.yml` - reduced CI suite, runs only 4x unit-tests (3.7-postgres, 3.8-mysql, 3.9-mysql, 3.10-postgres) instead of the full 8x set of unit tests, runs on any update to a pull request.

Also fixed a minor issue that I noticed in that our `selenium` container in CI was still set to the old 3.141 version instead of the 4.1.2 version that our local docker-compose file sets up.

# TODO
- [x] Do we even need 4x unit tests for each PR update, or should we reduce this even further to (3.7-postgres, 3.10-mysql) for greater efficiency?
- [x] Remove `on: pull_request` from `ci_integration.yml` before final merge.

### CODE REVIEW

**Comment by bryanculver** [comment]
> Do we even need 4x unit tests for each PR update, or should we reduce this even further to (3.7-postgres, 3.10-mysql) for greater efficiency?

Let's measure our improvement even with this alone. It's already much more streamlined than current, no need to over optimize. Where I currently stand is we should at least test all Python versions.

If I saw anything more minimal it would be one Python version, two DB into `develop` or `next`, exhaustive suite on PR to `main`.

**Comment by glennmatthews** [comment]
> If I saw anything more minimal it would be one Python version, two DB into `develop` or `next`, exhaustive suite on PR to `main`.

PRs to `main` should always and only ever come from `develop`, which will be running the full CI suite on each commit.

**Comment by jathanism** [comment]
Hmm, for `ci_pullrequest` it seems like it might be a good time a good time to introduce `concurrency`? See: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency

Specifically coming to a decision about cancelling jobs in progress, which would be another timesaver. For example to cancel in-progress jobs/runs for the current workflow:

```yaml
concurrency: 
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true
```

**Comment by glennmatthews** [comment]
Looks like `concurrency` works. Would it make sense to add to `ci_integration` as well, or do we want CI to run to completion on every push to the integration branches even if there are multiple in short order?

**Comment by bryanculver** [comment]
> Looks like `concurrency` works. Would it make sense to add to `ci_integration` as well, or do we want CI to run to completion on every push to the integration branches even if there are multiple in short order?

Depends on the visual output I guess. If it shows ❌ on cancelled runs then I say we leave concurrency out. I’d rather it be slow and complete with visual indicators as such.

**Comment by glennmatthews** [comment]
> > Looks like `concurrency` works. Would it make sense to add to `ci_integration` as well, or do we want CI to run to completion on every push to the integration branches even if there are multiple in short order?
> 
> Depends on the visual output I guess. If it shows ❌ on cancelled runs then I say we leave concurrency out. I’d rather it be slow and complete with visual indicators as such.

As you can see above it does show :x: on cancelled runs. Are we OK with that for pull requests at least?

**Comment by bryanculver** [comment]
> > > Looks like `concurrency` works. Would it make sense to add to `ci_integration` as well, or do we want CI to run to completion on every push to the integration branches even if there are multiple in short order?
> > 
> > 
> > Depends on the visual output I guess. If it shows ❌ on cancelled runs then I say we leave concurrency out. I’d rather it be slow and complete with visual indicators as such.
> 
> As you can see above it does show ❌ on cancelled runs. Are we OK with that for pull requests at least?

I say yes, as we’ll squash these commits and these branches will see a lot more very quickly.

File: .github/workflows/ci_integration.yml [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
+name: "CI - Integration Branch"
 on:  # yamllint disable
+  pull_request:  # TODO: remove this once verified
   push:
-  pull_request:
+    branches:
+      - "main"
+      - "develop"
+      - "next"
     paths-ignore:
```
Comment (glennmatthews) [inline .github/workflows/ci_integration.yml:11]: Does having `paths-ignore` still make sense for `ci_integration.yml`? Or should we have integration CI run on *every* commit to an integration branch, even those that only touch (for example) documentation files?

File: .github/workflows/ci_pullrequest.yml [hunk 2]
```diff
@@ -0,0 +1,234 @@
+---
+# A lighter-weight version of CI specifically for pull requests
+name: "CI - Pull Request"
+concurrency:  # Cancel any existing runs of this workflow for this same PR
+  group: "${{ github.workflow }}-${{ github.ref }}"
+  cancel-in-progress: true
+on:  # yamllint disable
+  pull_request:
+    paths-ignore:
```
Comment (glennmatthews) [inline .github/workflows/ci_pullrequest.yml:9]: Similar question here - with a lighter-weight ci_pullrequest, do we still want there to be files/commits that don't trigger a rerun of ci?

Comment (jathanism) [reply .github/workflows/ci_pullrequest.yml:9]: IMO let's pull it out and see how it goes. We can always add it back.

File: .github/workflows/ci_integration.yml (hunk 1 above)
Comment (jathanism) [reply .github/workflows/ci_integration.yml:10]: I think we're still left with the issue that without `paths-ignore`, there's temptation to use `[skip ci]` on commits that are documentation updates only, leaving the `integration-test` check dangling.

So long as such commits aren't the last commit in a PR branch going into an integration branch, the problem is moot, but we have to assert this somehow most likely in our release management "ritual".

File: .github/workflows/ci_pullrequest.yml (hunk 2 above)
Comment (bryanculver) [reply .github/workflows/ci_pullrequest.yml:9]: Agreed

File: .github/workflows/ci_integration.yml (hunk 1 above)
Comment (bryanculver) [reply .github/workflows/ci_integration.yml:10]: I think with the improvements made in this PR in general `[skip ci]` should not need to be used if much at all, especially with sights on #1635

**Review by bryanculver — APPROVED** [review]

**Pull request merged**